src/sync.cpp

```diff
@@ -135,16 +139,53 @@ static void potential_deadlock_detected(const LockPair& mismatch, const LockStac
     throw std::logic_error(strprintf("potential deadlock detected: %s -> %s -> %s", mutex_b, mutex_a, mutex_b));
 }
 
-static void push_lock(void* c, const CLockLocation& locklocation)
+static void double_lock_detected(const void* mutex, LockStack& lock_stack)
 {
+    LogPrintf("DOUBLE LOCK DETECTED\n");
+    LogPrintf("Lock order:\n");
+    for (const LockStackItem& i : lock_stack) {
+        if (i.first == mutex) {
+            LogPrintf(" (*)"); /* Continued */
+        }
+        LogPrintf(" %s\n", i.second.ToString());
+    }
+    // We throw or abort() here, never actually locking the mutex. So remove the entry
+    // from the lock stack which we just added in push_lock(). Tests in sync_tests.cpp
+    // continue executing past the exception and we don't want one test to leave a
+    // bricked stack for a subsequent test that could execute in the same thread.
+    lock_stack.pop_back();
```

Comment:
Would be nice if the `potential_deadlock_detected` and `double_lock_detected` helpers used the same style to pop_back the just added lock. (Either pass in a mutable reference or create a copy)

Reply:
Done - used the same as used when calling `potential_deadlock_detected()` in `master` - a copy.